### PR TITLE
User can buy a subscription

### PR DESCRIPTION
**Pull request by kaylawoodbury**
# [PT Story](https://www.pivotaltracker.com/story/show/171811135)
## User Story
```
As a user 
In order to get access to the best articles
I would like to be able to buy a subscription
```
## Changes proposed in this pull request:
* Sets up tests for user can buy subscription
* Adds button on premium articles for user to buy a subscription
* Adds logic to only show button if user is logged in
* Adds subscription form to collect user payment details
* Adds flash message to inform user if purchase was successful
## What I have learned working on this feature:
* How to set up an integration with Stripe
* How to test with Stripe
## Packages/Gems added
- react-stripe-elements
## Screenshots (Only if client side story)
Screenshot 2020-04-01 at 11 58 26
Screenshot 2020-04-01 at 11 58 32

## Lo-Fi (Only if client side story)
Pasted Image at 2020_03_17_08_26 am

### CODE REVIEW

File: src/components/SubscriptionForm.jsx [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            <label>CVC:</label>
+            <CardCVCElement />
+            <Button
+              margin="xsmall"
+              onClick={event => {
+                submitPayment(event);
+              }}
+            >
+              Submit Payment
+        </Button>
```
Comment (blake-futchi) [inline src/components/SubscriptionForm.jsx]: format indent

File: src/components/DisplaySingleArticle.jsx [hunk 2]
```diff
@@ -5,16 +5,29 @@ import { Button, Header, Icon, Image, Container } from "semantic-ui-react";
...
+  if (premiumUser === false && articlePremium === true && userEmail !== undefined) {
     article.content = article.content.substring(0, 200) + "...";
     premiumMessage = (
       <p id="premium-message">
-        <a href="">This article requires a premium membership.</a>
+        <a>This article requires a premium membership.</a>
+        <Button positive
+          onClick={ () => props.dispatch({type: 'SHOW_SUBSCRIPTION', payload: {showSubscription: true}})}
+          >Buy Subscription
+          </Button>
```
Comment (blake-futchi) [inline src/components/DisplaySingleArticle.jsx]: format indent

File: src/components/DisplaySingleArticle.jsx [hunk 3]
```diff
@@ -5,16 +5,29 @@ import { Button, Header, Icon, Image, Container } from "semantic-ui-react";
...
   let article = props.singleArticle;
   let premiumMessage = "";
 
-  if (premiumUser === false && articlePremium === true) {
+  if (premiumUser === false && articlePremium === true && userEmail !== undefined) {
     article.content = article.content.substring(0, 200) + "...";
     premiumMessage = (
       <p id="premium-message">
-        <a href="">This article requires a premium membership.</a>
+        <a>This article requires a premium membership.</a>
```
Comment (robin-lillqvist) [inline src/components/DisplaySingleArticle.jsx]: doesn't need to be <a> if we aren't supposed to click on it

File: src/components/DisplaySingleArticle.jsx [hunk 4]
```diff
@@ -5,16 +5,29 @@ import { Button, Header, Icon, Image, Container } from "semantic-ui-react";
...
+          >Buy Subscription
+          </Button>
+      </p>
+    );
+  }
+  else if (userEmail == undefined && articlePremium === true) {
+    article.content = article.content.substring(0, 200) + "...";
+    premiumMessage = (
+      <p id="premium-message">
+        <a>This article requires a premium membership. Please login to purchase a subscription.</a>
```
Comment (robin-lillqvist) [inline src/components/DisplaySingleArticle.jsx]: doesn't need to be if we aren't supposed to click on it

File: src/components/DisplaySingleArticle.jsx (hunk 3 above)
Comment (blake-futchi) [inline src/components/DisplaySingleArticle.jsx]: make it p or h1/h2 etc

File: src/components/DisplaySingleArticle.jsx (hunk 4 above)
Comment (blake-futchi) [inline src/components/DisplaySingleArticle.jsx]: remove a tag and make this p or something else

File: src/components/SubscriptionForm.jsx [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    let paymentStatus = await axios.post("/subscriptions", {
+      stripeToken: token,
+      email: userEmail
+    },
+      { headers: headers }
+    );
+    if (paymentStatus.data.status === "paid")
+    dispatch({
+      type: "FLASH_MESSAGE",
+      payload: { flashMessage: "Thank you for your business!", showArticlesList: true, showSubscription: false, premiumUser: true },
```
Comment (robin-lillqvist) [inline src/components/SubscriptionForm.jsx]: Change to something more personal here.

File: src/components/SubscriptionForm.jsx [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+  return (
+    <>
+      <Segment raised compact>
+        <Form id="payment-form">
+          <Header textAlign="center" as='h2' dividing>
+            Payment Form
+        </Header>
+          <Header textAlign="center" as='h5'>
+            Step above the crowd, with our Premium Platinum Plan costing the low amount of 10,000SEK per year.
```
Comment (blake-futchi) [inline src/components/SubscriptionForm.jsx]: "costing the low amount" sounds very non native english... Plan for only 10,000 maybe

File: src/state/store/initialState.js [hunk 7]
```diff
@@ -9,7 +9,8 @@ const initialState = {
   authenticate: false,
   userEmail: undefined,
   message: "",
-  premiumUser: false
+  premiumUser: false,
+  flashMessage: ''
```
Comment (blake-futchi) [inline src/state/store/initialState.js]: consistency on ' ' vs " "

**Review by blake-futchi — COMMENTED** [review]
there is a file called { , why is this there?

**Review by robin-lillqvist — COMMENTED** [review]
Comments

**Pull request merged**